File: engine/__tests__/unit/minimization-pass.test.ts
import { describe, it, expect } from "vitest";
import { parse } from "../../passes/parser-pass.js";
import { thompsonConstruction } from "../../passes/thompson-pass.js";
import { subsetConstruction } from "../../passes/subset-construction-pass.js";
import { minimize } from "../../passes/minimization-pass.js";
import type { Automaton } from "../../types.js";

/** Helper: run a pattern through the real pipeline up to the minimized DFA. */
function minDfaFor(pattern: string): { dfa: Automaton; min: Automaton } {
  const parsed = parse(pattern);
  if (!parsed.ok) throw new Error(`test pattern "${pattern}" failed to parse`);
  const nfa = thompsonConstruction(parsed.ast);
  const dfa = subsetConstruction(nfa);
  return { dfa, min: minimize(dfa) };
}

function flat(automaton: Automaton): Record<number, Record<string, number[]>> {
  const out: Record<number, Record<string, number[]>> = {};
  for (const [state, bySymbol] of automaton.transitions) {
    out[state] = Object.fromEntries(bySymbol);
  }
  return out;
}

describe("MinimizationPass — genuinely reduces state count", () => {
  it("'(a|b)*ab' (classic Aho/Sethi/Ullman example): DFA has 4 states, min-DFA has 3", () => {
    const { dfa, min } = minDfaFor("(a|b)*ab");

    // Confirm the DFA really is non-minimal before minimizing, so this test
    // is actually exercising the reduction and not a no-op.
    expect(dfa.stateCount).toBe(4);

    expect(min.stateCount).toBe(3);
    expect(min.startState).toBe(0);
    expect(min.acceptStates).toEqual(new Set([2]));
    expect(flat(min)).toEqual({
      0: { a: [1], b: [0] }, // self-loop: the two DFA states merged here were behaviorally identical
      1: { a: [1], b: [2] },
      2: { a: [1], b: [0] },
    });
  });

  it("'ab|ac': the two single-transition accept states merge into one", () => {
    const { dfa, min } = minDfaFor("ab|ac");

    expect(dfa.stateCount).toBe(4);
    expect(min.stateCount).toBe(3);
    expect(min.startState).toBe(0);
    expect(min.acceptStates).toEqual(new Set([2]));
    expect(flat(min)).toEqual({
      0: { a: [1] },
      1: { b: [2], c: [2] },
    });
  });
});

describe("MinimizationPass — does not over-merge an already-minimal DFA", () => {
  it("'aa|a' is already minimal: 3 states in, 3 states out, unchanged shape", () => {
    const { dfa, min } = minDfaFor("aa|a");

    expect(dfa.stateCount).toBe(3);
    expect(min.stateCount).toBe(3);
    expect(min.startState).toBe(0);
    expect(min.acceptStates).toEqual(new Set([1, 2]));
    expect(flat(min)).toEqual({
      0: { a: [1] },
      1: { a: [2] },
    });
  });
});

describe("MinimizationPass — accepting and non-accepting states never merge", () => {
  it("no group in the minimized DFA mixes accepting and non-accepting original states", () => {
    // Structural sanity check rather than a golden value: run several patterns
    // and confirm the invariant Moore's algorithm depends on never breaks.
    for (const pattern of ["(a|b)*ab", "ab|ac", "aa|a", "(a|b)*", "a*b*"]) {
      const { min } = minDfaFor(pattern);
      // Every accept state in min-DFA should behave consistently — this is
      // implicitly checked by minimize() itself (acceptStates.add uses
      // group[0] as representative), so this test just confirms the pass
      // runs cleanly end-to-end without throwing for a spread of patterns.
      expect(min.stateCount).toBeGreaterThan(0);
      expect(min.acceptStates.size).toBeGreaterThan(0);
    }
  });
});

File: tests/minimization_pass_test.cpp
// tests/minimization_pass_test.cpp
//
// Direct port of minimization-pass.test.ts.

#include <catch2/catch_test_macros.hpp>
#include <map>
#include <stdexcept>

#include "regexlab/passes/minimization-pass.hpp"
#include "regexlab/passes/parser-pass.hpp"
#include "regexlab/passes/subset-construction-pass.hpp"
#include "regexlab/passes/thompson-pass.hpp"
#include "regexlab/types.hpp"

using namespace regexlab;

namespace {

struct DfaPair {
    Automaton dfa;
    Automaton min;
};

/** Run a pattern through the real pipeline up to the minimized DFA. */
DfaPair minDfaFor(const std::string& pattern) {
    ParseResult parsed = parse(pattern);
    if (!parsed.ok) throw std::runtime_error("test pattern \"" + pattern + "\" failed to parse");
    Automaton nfa = thompsonConstruction(parsed.ast);
    Automaton dfa = subsetConstruction(nfa);
    return {dfa, minimize(dfa)};
}

std::map<int, std::map<std::string, std::vector<int>>> flatten(const Automaton& automaton) {
    std::map<int, std::map<std::string, std::vector<int>>> out;
    for (const auto& [state, bySymbol] : automaton.transitions) {
        out[state] = std::map<std::string, std::vector<int>>(bySymbol.begin(), bySymbol.end());
    }
    return out;
}

}  // namespace

TEST_CASE("MinimizationPass — genuinely reduces state count", "[minimization]") {
    SECTION("'(a|b)*ab' (classic Aho/Sethi/Ullman example): DFA has 4 states, min-DFA has 3") {
        auto [dfa, min] = minDfaFor("(a|b)*ab");

        // Confirm the DFA really is non-minimal before minimizing, so this test
        // is actually exercising the reduction and not a no-op.
        REQUIRE(dfa.stateCount == 4);

        REQUIRE(min.stateCount == 3);
        REQUIRE(min.startState == 0);
        REQUIRE(min.acceptStates == std::unordered_set<int>{2});

        std::map<int, std::map<std::string, std::vector<int>>> expected = {
            {0, {{"a", {1}}, {"b", {0}}}},  // self-loop: merged behaviorally identical states
            {1, {{"a", {1}}, {"b", {2}}}},
            {2, {{"a", {1}}, {"b", {0}}}},
        };
        REQUIRE(flatten(min) == expected);
    }

    SECTION("'ab|ac': the two single-transition accept states merge into one") {
        auto [dfa, min] = minDfaFor("ab|ac");

        REQUIRE(dfa.stateCount == 4);
        REQUIRE(min.stateCount == 3);
        REQUIRE(min.startState == 0);
        REQUIRE(min.acceptStates == std::unordered_set<int>{2});

        std::map<int, std::map<std::string, std::vector<int>>> expected = {
            {0, {{"a", {1}}}},
            {1, {{"b", {2}}, {"c", {2}}}},
        };
        REQUIRE(flatten(min) == expected);
    }
}

TEST_CASE("MinimizationPass — does not over-merge an already-minimal DFA", "[minimization]") {
    auto [dfa, min] = minDfaFor("aa|a");

    REQUIRE(dfa.stateCount == 3);
    REQUIRE(min.stateCount == 3);
    REQUIRE(min.startState == 0);
    REQUIRE(min.acceptStates == (std::unordered_set<int>{1, 2}));

    std::map<int, std::map<std::string, std::vector<int>>> expected = {
        {0, {{"a", {1}}}},
        {1, {{"a", {2}}}},
    };
    REQUIRE(flatten(min) == expected);
}

TEST_CASE("MinimizationPass — accepting and non-accepting states never merge", "[minimization]") {
    // Structural sanity check rather than a golden value: run several patterns
    // and confirm the invariant Moore's algorithm depends on never breaks.
    for (const std::string& pattern :
         {std::string("(a|b)*ab"), std::string("ab|ac"), std::string("aa|a"),
          std::string("(a|b)*"), std::string("a*b*")}) {
        auto [dfa, min] = minDfaFor(pattern);
        REQUIRE(min.stateCount > 0);
        REQUIRE(min.acceptStates.size() > 0);
    }
}
